File: BACK/src/main/java/com/genesis/api/statmanager/service/JoueurService.java
package com.genesis.api.statmanager.service;

import com.genesis.api.statmanager.dto.global.StatistiquesDTO;
import com.genesis.api.statmanager.dto.joueur.JoueurDTO;
import com.genesis.api.statmanager.dto.joueur.JoueurLightDTO;
import com.genesis.api.statmanager.dto.joueur.JoueurStatGlobalDTO;
import com.genesis.api.statmanager.dto.statistique.StatCompositeDTO;
import com.genesis.api.statmanager.dto.statistique.StatCritereDTO;
import com.genesis.api.statmanager.dto.statistique.StatTopJoueurDTO;

import java.util.List;

/**
 * 📌 Service de gestion des joueurs.
 */
public interface JoueurService {

    /**
     * 📌 Récupère tous les joueurs sous forme de DTO léger.
     *
     * @return Liste de JoueurDTO.
     */
    List<JoueurDTO> findAllAsDTO();

    /**
     * 📌 Récupère un joueur précis (DTO) par ID.
     *
     * @param id ID du joueur.
     * @return JoueurDTO.
     */
    JoueurDTO findByIdAsDTO(Long id);

    /**
     * 📌 Récupère une liste allégée des joueurs (ID, Nom, Poste) triée par poste.
     *
     * @return Liste de JoueurDTO (light).
     */
    List<JoueurLightDTO> getJoueursLight();

    /**
     * 📌 Récupère un joueur avec ses statistiques globales (buts, passes, minutes, cote...).
     *
     * @param id ID du joueur.
     * @return JoueurStatGlobalDTO.
     */
    JoueurStatGlobalDTO findByIdWithGlobalStats(Long id);

    /**
     * 📌 Récupère tous les joueurs avec leurs statistiques globales.
     *
     * @return Liste de JoueurStatGlobalDTO.
     */
    List<JoueurStatGlobalDTO> findAllWithGlobalStats();

    /**
     * 📌 Classe et retourne les meilleurs joueurs selon un critère (ex: buts, passes) et/ou par poste.
     *
     * @param critere Critère de tri (ex: "buts", "passes", "scoreOffensifGlobal", etc.).
     * @param poste   Filtre sur un poste spécifique (ex: "AC"), ou null pour ne pas filtrer.
     * @return Liste (max 10) des meilleurs joueurs correspondant au critère.
     */
    List<JoueurStatGlobalDTO> getTopJoueursGlobaux(String critere, String poste);

    /**
     * 📌 Retourne les joueurs triés selon un critère spécifique.
     */
    StatCritereDTO getJoueursParCritere(String critere);

    /**
     * 📌 Retourne les 5 meilleurs joueurs des dernières rencontres.
     */
    StatTopJoueurDTO getMeilleursJoueurs();

    /**
     * 📌 Retourne les joueurs triés selon un score composite.
     */
    StatCompositeDTO getJoueursParScoreComposite();

    List<StatistiquesDTO> getAllJoueursStats();

    JoueurStatGlobalDTO findByNomWithGlobalStats(String nom);
}

File: front/src/app/modules/joueur/services/joueur.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from './../../../../environments/environment'; // ✅ Import de l'environnement
// ► Import des interfaces ou classes que tu vas utiliser
import { JoueurDTO } from '../models/joueur.model';
import { JoueurLightDTO, JoueurStatGlobalDTO } from '../models/joueur.model'; // Par ex.
import { StatistiquesDTO } from './../../statistiques/models/statistiques.model';

@Injectable({
  providedIn: 'root',
})
export class JoueurService {
  private apiUrl = `${environment.apiBaseUrl}/joueur`;

  constructor(private http: HttpClient) {}

  /**
   * Récupère la version "light" des joueurs (ID, Nom, Poste).
   * GET /joueur/light
   */
  getJoueursLight(): Observable<JoueurLightDTO[]> {
    return this.http.get<JoueurDTO[]>(`${this.apiUrl}/light`);
  }

  /**
   * Récupère la version "complète" de tous les joueurs.
   * GET /joueur/all
   */
  getJoueurs(): Observable<JoueurDTO[]> {
    return this.http.get<JoueurDTO[]>(`${this.apiUrl}/all`);
  }

  /**
   * 📌 Récupère tous les joueurs avec leurs statistiques globales.
   * GET /api/joueur/stats/all
   */
  getAllJoueursStats(): Observable<StatistiquesDTO[]> {
    return this.http.get<StatistiquesDTO[]>(`${this.apiUrl}/all`);
  }

  /** 📌 Récupère les statistiques globales d'un joueur */
  getJoueurStats(idJoueur: number): Observable<JoueurStatGlobalDTO> {
    return this.http.get<JoueurStatGlobalDTO>(`${this.apiUrl}/${idJoueur}/stats`);
  }

  // 📌 Récupérer un joueur via son Nom
  getJoueurStatsByNom(nom: string): Observable<JoueurStatGlobalDTO> {
    return this.http.get<JoueurStatGlobalDTO>(`${this.apiUrl}/nom/${nom}/stats`);
  }
}
